pkg/tms/tmsUtils.go

```diff
@@ -140,6 +146,14 @@ func unmarshalServiceKey(serviceKeyJson string) (serviceKey serviceKey, err erro
 	if err != nil {
 		return
 	}
+	if len(serviceKey.Uri) == 0 {
+		if serviceKey.CALMEndpoints != nil && len(serviceKey.CALMEndpoints.API) > 0 {
+			serviceKey.Uri = serviceKey.CALMEndpoints.API + CALM_REROUTING_ENDPOINT_TO_CTMS
+		} else {
+			err = fmt.Errorf("neither uri nor endpoints.Api is set in service key json string")
+			return
+		}
+	}
```

Comment:
[minor] With regards to logging in line 256: although the parameter is still called tmsServiceKey, the service key being parsed might be now a cALM one. Does it make sense to adjust the logging message in line 256  for being more generic? E.g. "Failed to unmarshal value of 'tmsServiceKey' parameter".

Reply:
I would propose changing this parameter first to be called simply serviceKey, what do you think about that @artembannikov @gerstneralex ?
Alternatively I will at least update the tmsServiceKey description to explain that also calm service keys are valid here.

Reply:
> I would propose changing this parameter first to be called simply serviceKey

Since the tmsServiceKey is already used in the Customers' yaml configurations, renaming would mean either going through the entire deprecation process (i.e. first supporting both of the parameters and then removing the tmsServiceKey after some time). But even if correctly organized the deprecation cannot prevent from breaking some Customers' pipelines, when the tmsServiceKey parameter is completely removed. Or it means "permanent" implicit support of the old tmsServiceKey parameter.

Reply:
Very true. Then our options are:
1. Introducing a new parameter: This would mean more effort on code side handling this.
2. Documenting this. 

I will go for 2 if there are no objections.

Reply:
@o-liver, I guess, under documentig you mean mentioning that tmsServiceKey parameter can also accept cALM service key as its value. Honestly saying, I do not like it so much - such remarks are always confusing. But eventhough I admit that the effort for introducing another service key parameter will be at least not small, I would personally go for 1, cause it is a cleaner solution.

@gerstneralex, what is your view on that?

Reply:
I will investigate the effort of introducing another serviceKey parameter, sure. 
Would you prefer a new service key called simply `serviceKey,` which can be either tms or calm service key? With the idea of deprecating the `tmsServiceKey` at some point.
Or make it explicitly `calmServiceKey?`

Reply:
There is also an fallback with the parameter "credentialsid" (https://www.project-piper.io/steps/tmsExport/#tmsservicekey). If we introduced the new calmServiceKey parameter how would we handle that? It think it's an additional argument for the option 2.

Reply:
> There is also an fallback with the parameter "credentialsid" (https://www.project-piper.io/steps/tmsExport/#tmsservicekey). If we introduced the new calmServiceKey parameter how would we handle that?

Yes, correct. Then the only clean way would be to introduce the new serviceKey parameter that can handle either cTMS or cALM key with further deprecation of tmsServiceKey parameter. Disadvantage: we will need to go through the proper deprecation process; on complete deprecation, it is not guaranteed that we will not break Customers' pipelines, or better to say, most probably, we will break some pipelines and affected Customers will need to adjust them.

Reply:
How about using the Jenkins `credentialsId` to our advantage:
1. We introduce a new parameter `serviceKey` that can handle both and deprecate the old `tmsServiceKey` with warnings in the documentation and console log
2. We introduce a new Jenkins `tmsCredentialyId` that we route to the `tmsServiceKey`
3. We route the existing Jenkins `credentialsId` to the new `serviceKey` parameter that can handle both.

What do you think @artembannikov @gerstneralex ?

Reply:
Or, better yet. We don't offer a Jenkins credential at all that is routed to tmsServiceKey parameter. We just deprecate the tmsServiceKey parameter and route the Jenkins `credentialsId` to the new `serviceKey` parameter that can handle both.

Reply:
Please have a look at the metadata/tms*.yaml files. Code changes are coming up.

Reply:
> Or, better yet. We don't offer a Jenkins credential at all that is routed to tmsServiceKey parameter. We just deprecate the tmsServiceKey parameter and route the Jenkins `credentialsId` to the new `serviceKey` parameter that can handle both.

Yes, exactly. The secret stored in Jenkins under the name provided as value of `credentailsId` parameter is a service key. Does not matter, which key is provided there, meaning cTMS or cALM. It will be obtained and parsed (also using your additional parsing logic for cALM keys). So the name of this parameter should stay service agnostic. The same as the name of the `serviceKey` parameter that we plan to introduce.

Reply:
With this commit the code is also implemented: 473d004f66ca75f80135e159320818467f2990b1
This was very easy as the config is transformed to this `tms.Options` object, wonderful!

Reply:
Unit tests are done